include/boost/http_proto/service/zlib_service.hpp

```diff
@@ -54,60 +50,143 @@ encoding_size_hint(decoder_config cfg = {}) noexcept
+struct stream
+{
+    /** Call the underling compression/decompression algorithm.
+
+        @param p The input and output buffers.
+
+        @param f The flush method.
+
+        @return The result of operation that contains a value
+        of @ref error.
+    */
+    virtual system::error_code
+    write(params& p, flush f) noexcept = 0;
+};
+
+/** Provides in-memory compression and decompression functions
+    using zlib underneath.
+*/
 struct BOOST_HTTP_PROTO_ZLIB_DECL
     service
     : http_proto::service
 {
-    struct stream
-    {
-        enum class flush
-        {
-            none,
-            partial,
-            sync,
-            full,
-            finish,
-            block,
-            trees
-        };
-
-        struct results
-        {
-            std::size_t out_bytes;
-            std::size_t in_bytes;
-            system::error_code ec;
-            bool finished;
-        };
-
-        virtual results
-        write(
-            buffers::mutable_buffer out,
-            buffers::const_buffer in,
-            flush) noexcept = 0;
-    };
-
     virtual
     std::size_t
     space_needed() const noexcept = 0;
 
+    /** Create a deflator stream by calling zlib `deflateInit2()`.
+
+        @param ws A reference to the workspace used for constructing the
+        deflator stream object and for storage used by zlib.
+
+        @param level The compression level.
+
+        @param window_bits The window size.
+
+        @param mem_level Specifies how much memory should be allocated
+        for the internal compression state.
+
+        @return A reference to the created deflator stream.
+
+        @throws std::length_error If there is insufficient free space in 
+        @ref `http_proto::detail::workspace`.
+    */
     virtual stream&
     make_deflator(
         http_proto::detail::workspace& ws,
         int level,
         int window_bits,
         int mem_level) const = 0;
 
+    /** Create an inflator stream by calling zlib `inflateInit2()`.
+
+        @param ws A reference to the workspace used for constructing the
+        inflator stream object and for storage used by zlib.
+
+        @param window_bits The window size.
+
+        @return A reference to the created inflator stream.
+
+        @throws std::length_error If there is insufficient free space in
+        @ref `http_proto::detail::workspace`.
+    */
     virtual stream&
     make_inflator(
         http_proto::detail::workspace& ws,
         int window_bits) const = 0;
 };
 
-void BOOST_HTTP_PROTO_ZLIB_DECL
+
+/** Installs a zlib service on the provided context.
+
+    @param ctx A reference to the @ref context where the service
+    will be installed.
+
+    @throw std::invalid_argument If the zlib service already
+    exist on the context.
+*/
+BOOST_HTTP_PROTO_ZLIB_DECL
+void
 install_service(context& ctx);
 
 } // zlib
 } // http_proto
 } // boost
 
+#include <boost/http_proto/service/impl/zlib_service.hpp>
```

Comment:
we don't do this anymore... everything goes in the compiled lib

Reply:
The contents of `impl/zlib_service.hpp` are declarations and constexpr functions, which must be placed in header files.
This is similar to: https://github.com/cppalliance/http_proto/blob/develop/include/boost/http_proto/impl/error.hpp